src/stockpile.clj

```diff
@@ -0,0 +1,364 @@
+    (durably-establish (.resolve top "stockpile")
+                       ;; Assumes that copy won't use a
+                       ;; BufferedWriter (FilterOutputStream) in this
+                       ;; case (otherwise it'll be broken with at
+                       ;; least openjdk-7.
+                       #(with-open [out (FileOutputStream. (.toFile %))]
+                          (.write out (.getBytes "0 stockpile" "UTF-8")))
+                       false)
+    (fsync top true)
+    {:directory top
+     :encode-meta encode-meta
+     :next-likely-id (AtomicLong. 0)}))
+
+(defn open
+  "Opens the queue in directory, and returns the queue.
+  Calls existing-entry-reducer as-per reduce for each existing entry
+  in q (the ordering of the calls is unspecified).  The reduction may
+  be escaped by throwing a unique exception (cf. slingshot).  Returns
+  [q reduction].  For example: (open \"foo\" conj []).  Calls
+  (encode-meta metadata-string) and (decode-meta encoded-string) to
+  handle encoding and decoding entry metadata to and from filesystem
+  Path strings."
+  [directory existing-entry-reducer reduction-base-val
+   & {:keys [encode-meta decode-meta]
+      :or {encode-meta encode-portable-path-element
+           decode-meta decode-portable-path-element}}]
+  (let [top (as-path directory)
+        q (.resolve top "q")]
+    (let [info-file (.resolve top "stockpile")
+          info (String. (Files/readAllBytes info-file) "UTF-8")]
+      (when-not (= "0 stockpile" info)
+        (throw (IllegalStateException.
+                (format "Invalid queue token %s found in %s"
+                        (pr-str info)
+                        (pr-str info-file))))))
+    (let [[q reduction] (fs->queue top decode-meta
+                                   existing-entry-reducer reduction-base-val)]
+      [(assoc q :encode-meta encode-meta) reduction])))
+
+(defn store
+  "Atomically and durably enqueues the content of stream, and returns
+  an entry that can be used to refer to the content later.  If an
+  ex-info exception of {:kind ::commit-failed} is thrown, it's
+  possible that the metadata was too long for the underlying
+  filesystem.  If provided, (adjust-metadata retry-count
+  previous-metadata) will be called whenever there is a failure that
+  might be related to the length of the metadata string.  It can
+  either return an alternate string, or nil/false to give up, which
+  will cause the ::commit-failed exception."
+  ([q stream] (store q stream nil))
+  ([q stream metadata] (store q stream metadata nil))
+  ([q stream metadata adjust-metadata]
+   (let [likely-id (.getAndIncrement (:next-likely-id q))]
+     (when-not likely-id
+       (throw (IllegalStateException.
+               (format "cannot write to queue in %s while opening it"
+                       (pr-str (-> (:directory q) .toAbsolutePath str))))))
+     (let [tmp-dest (create-tmp-file (.resolve (:directory q) "q"))
+           encode-meta (:encode-meta q)]
```

Comment:
I'm ok with that being a known issue, users just shouldn't do that. If the user was to hit that, their workaround is easy. Startup in the original locale, process all messages, shutdown. Change the locale and start back up. Right?